File: web/src/storyAudio.ts
/**
 * Hearing a story: the third place in Acervo where anything is heard, after `pronunciation.ts` (a word
 * or a sentence) and `loops.ts` (a track), and modelled on the second.
 *
 * **A part is a queue of passages, and nothing is ever seeked.** Each passage is its own file, so
 * playing one is playing it from its first sample, and playing from a passage to the end of the part
 * is playing that file and then the ones after it. The alternative was tried and does not work: join
 * the passages, remember where each begins, and set `currentTime`. A browser seeks a compressed
 * stream to a page boundary — one second, in the Ogg the server writes — so a passage started a word
 * or two late, or a word or two into the sentence before, unpredictably; and the element reports the
 * position that was *asked for* rather than the one it gave, so nothing can detect it or correct it.
 * A file that begins where the passage begins cannot be wrong.
 *
 * **Pause is pause, and leaving is stop.** Pressing play on a part that is playing pauses it, and
 * pressing it again carries on from where it was. There is no scrubber — a story is read a part at a
 * time, and swiping to another part and back is the way to begin again, because leaving a part
 * *stops* it and drops its position. `stopPart` is what the reader calls when a page stops being the
 * one on screen.
 *
 * **A passage is a place to start.** Touching one when nothing of this part is playing plays that
 * passage alone; touching one while the part is playing goes there and carries on to the end. The
 * first answers "what was that sentence?", the second "go back to there".
 *
 * **On demand is the same route the job takes.** A part with no passages asks the server for them,
 * which is one call to the same function `story.audio` runs, and plays when they arrive. The element
 * is started on silence in the press itself, before that wait, for `pronunciation.ts`'s reason: iOS
 * lets an element play only from inside the gesture that asked.
 *
 * One sound at a time: this stops `pronunciation.ts`, pauses every other registered player before it
 * plays, and registers its own pause with them. The dependency runs one way, as it does for loops.
 */

import { useSyncExternalStore } from "react";
import { AcervoApiError, backendSession } from "./api";
import type { AudioSegment, StoryPart } from "./domain";
import { clipBlob, registerPlayer, silencePlayers, stop as stopSpeech } from "./pronunciation";
import { repository } from "./repository";
import { syncEngine } from "./sync";

export type StoryAudioStatus = "idle" | "busy" | "playing" | "paused";

export interface StoryPlayback {
  /** The part being fetched, recorded, heard or paused. Kept while `failed`, so its row can say why. */
  partId: string | null;
  status: StoryAudioStatus;
  /** Index of the passage that is sounding, or null when nothing of this part is. */
  segment: number | null;
  failed: string | null;
}

const EMPTY: StoryPlayback = { partId: null, status: "idle", segment: null, failed: null };

let state: StoryPlayback = EMPTY;
const listeners = new Set<() => void>();

function update(partial: Partial<StoryPlayback>): void {
  state = { ...state, ...partial };
  listeners.forEach((listener) => listener());
}

const subscribe = (listener: () => void) => { listeners.add(listener); return () => { listeners.delete(listener); }; };

export const storyPlayback = (): StoryPlayback => state;

export function useStoryAudio(): StoryPlayback {
  return useSyncExternalStore(subscribe, storyPlayback);
}

/* ── the element ─────────────────────────────────────────────────────── */

let element: HTMLAudioElement | null = null;
/** The object URL the element is holding, revoked when it is let go. */
let held: string | null = null;
/** Bumped by every press that starts something, so an answer that arrives late knows it is stale. */
let run = 0;

/** What is queued: the passages, which one is sounding, which is the last, and their bytes. */
let queue: { passages: AudioSegment[]; at: number; last: number; blobs: Map<string, Blob> } | null = null;

function audio(): HTMLAudioElement {
  if (element) return element;
  element = new Audio();
  element.preload = "auto";
  // The passage's own end only: the silence `prime` starts ends too, long before one has arrived.
  element.addEventListener("ended", () => { if (held && element?.src === held) void advance(); });
  element.addEventListener("pause", () => {
    if (state.status === "playing" && held && element?.src === held) update({ status: "paused" });
  });
  element.addEventListener("play", () => {
    if (held && element?.src === held) update({ status: "playing" });
  });
  return element;
}

/* A few milliseconds of silence, for the reason `pronunciation.ts` and `loops.ts` give. */
const SILENCE = "data:audio/wav;base64,UklGRiQAAABXQVZFZm10IBAAAAABAAEAQB8AAIA+AAACABAAZGF0YQAAAAA=";

function prime(): void {
  const player = audio();
  if (!player.paused) return;
  try { player.src = SILENCE; void player.play()?.catch(() => undefined); }
  catch { /* jsdom, or a browser with no audio at all */ }
}

/** Let go of the element: nothing is playing, nothing is held, and the queue is forgotten. */
function release(): void {
  if (element) {
    try { element.pause(); } catch { /* nothing to stop */ }
    element.removeAttribute("src");
  }
  if (held) { URL.revokeObjectURL(held); held = null; }
  queue = null;
}

/** Play the passage the queue is on. */
async function sound(): Promise<void> {
  if (!queue) return;
  const passage = queue.passages[queue.at];
  const blob = queue.blobs.get(passage.audioRef);
  if (!blob) { finish(); return; }
  const player = audio();
  if (held) URL.revokeObjectURL(held);
  held = URL.createObjectURL(blob);
  player.src = held;
  update({ status: "playing", segment: queue.at });
  await player.play();
}

/** This passage ended: on to the next, unless it was the last that was asked for. */
async function advance(): Promise<void> {
  if (!queue) return;
  if (queue.at >= queue.last) { finish(); return; }
  queue.at += 1;
  try {
    await sound();
  } catch (error) {
    console.warn("Acervo: a story passage could not be played", error);
    finish();
  }
}

function finish(): void {
  run += 1;
  release();
  update(EMPTY);
}

/* ── pressing ────────────────────────────────────────────────────────── */

function failure(error: unknown): string {
  if (error instanceof AcervoApiError && error.code === "offline") {
    return "Reading a story aloud needs the server, and it cannot be reached.";
  }
  if (error instanceof DOMException && error.name === "NotAllowedError") {
    return "This device would not start the audio. Press play again.";
  }
  return error instanceof Error ? error.message : "That part could not be read aloud.";
}

/**
 * Play `part` from passage `from`, stopping after `last`.
 *
 * Every passage that will be played is fetched before the first one sounds. They are a few tens of
 * kilobytes each and already on the device in the ordinary case, and holding them is what makes the
 * swap from one to the next an assignment rather than a fetch — which keeps the seam between two
 * sentences as short as the silence a speaker leaves anyway.
 */
async function start(part: StoryPart, from: number, last: number): Promise<void> {
  const mine = ++run;
  release();
  stopSpeech();
  silencePlayers(pause);
  // Before any await, so the browser still counts this as the gesture that asked.
  prime();
  update({ partId: part.id, status: "busy", segment: null, failed: null });
  try {
    let passages = part.audioSegments;
    if (!passages.length) {
      const row = await backendSession.recordStoryPart(part.storyId, part.id, repository.snapshot().deviceId);
      if (mine !== run) return;
      // The row reaches the replica the way every server-written row does; it plays meanwhile.
      void syncEngine.syncNow();
      passages = row.audioSegments;
      from = 0;
      last = passages.length - 1;
    }
    if (!passages.length) throw new Error("That part has no recording.");
    const stop = Math.min(last, passages.length - 1);
    const blobs = new Map<string, Blob>();
    for (const passage of passages.slice(from, stop + 1)) {
      blobs.set(passage.audioRef, await clipBlob(passage.audioRef));
      if (mine !== run) return;
    }
    queue = { passages, at: from, last: stop, blobs };
    await sound();
  } catch (error) {
    if (mine !== run) return;
    console.warn("Acervo: a story part could not be played", { part: part.id, error });
    release();
    update({ partId: part.id, status: "idle", segment: null, failed: failure(error) });
  }
}

/** Play, pause or carry on, for the button on a part. Pressing it while it is being made stops that. */
export function toggle(part: StoryPart): void {
  if (state.partId === part.id) {
    if (state.status === "playing") { pause(); return; }
    if (state.status === "paused") { resume(); return; }
    if (state.status === "busy") { stop(); return; }
  }
  void start(part, 0, Math.max(part.audioSegments.length - 1, 0));
}

/**
 * A touch on a passage. While this part is sounding it goes there and carries on to the end;
 * otherwise it plays that passage alone.
 */
export function playSegment(part: StoryPart, index: number): void {
  if (!part.audioSegments[index]) return;
  const sounding = state.partId === part.id && (state.status === "playing" || state.status === "paused");
  void start(part, index, sounding ? part.audioSegments.length - 1 : index);
}

export function pause(): void {
  if (state.status !== "playing") return;
  try { element?.pause(); } catch { /* nothing to stop */ }
  update({ status: "paused" });
}

function resume(): void {
  void audio().play().catch((error: unknown) => {
    release();
    update({ status: "idle", segment: null, failed: failure(error) });
  });
}

/** Stop and forget the position. */
export function stop(): void {
  run += 1;
  release();
  update(EMPTY);
}

/**
 * Stop, but only if it is this part that is being heard — what a page does when it stops being the
 * one on screen, without silencing a part that is somewhere else.
 */
export function stopPart(partId: string): void {
  if (state.partId === partId) stop();
}

export function replaceForTests(): void {
  run += 1;
  release();
  element = null;
  state = EMPTY;
}

registerPlayer(pause);

File: src/acervo/services/story_audio.py
"""Acervo's binding to story narration: a part in, one recording and its row out.

The other half of the split `acervo.stories.narrate` makes, and the sibling of `services/stories.py`
(the pictures) and `services/pronunciations.py` (a word). This is where it is decided *whose* voice
reads it, and it reuses what those two already answer: the chain a pair is chosen from, the way a
model's refusal is spoken, and the file-then-row write.

**A story is read in one voice, but the pin is a preference and not a cage.** The first part to be
recorded walks the owner's order, and whichever (provider, model, voice) answers is asked for every
later passage and part. If that pair then refuses — a daily quota is the case this exists for — the
next pair in the order records the rest instead. A story that changes voice between parts is worse
than one read throughout in a single voice and far better than one that stops after part one, which
is what insisting cost: the pinned pair was the free row whose allowance had run out, the chain of one
had nothing to fall through to, and three parts stayed silent for a day. **Within a part the pin never
moves**, so a paragraph is never read by two voices. The pin is read back off the graph — the first
part already recorded — and never remembered, which is what lets a retry, a route and a job agree
without coordinating.

**Only a voice that can take a direction reads a directed story.** The passages exist to carry one, so
a pair that cannot take one turns four calls a part into four times the cost of one, for nothing: on
the deployment this was found on, every stored passage read `direction=""` because the row at the head
of the order goes through LiteLLM, whose Vertex speech transformation has no field for a prompt at all.
So the expressive order is filtered to the pairs that declare `style: instruction`, and **when none of
them can be reached the part is read whole, in one call, with no passages** — a plain reading of the
whole paragraph rather than an expensive one that is plain anyway.

**A part is one file per passage, and never one joined file.** A directed voice is asked for one
passage at a time, each with its own direction (`stories/narrate.py` decides where they break), and
each recording is stored as it came. The alternative was tried: join them, remember where each one
starts, and seek. It does not work, because **a browser seeks a compressed stream to a page boundary**
— one second, in the Ogg libsndfile writes — so a passage started after its first words or after the
end of the one before it, unpredictably, and the element reports the time that was *asked for* rather
than the time it gave, so there is nothing to correct against. A file that begins where the passage
begins needs no seek and cannot be wrong. A clear voice records one passage covering the whole part,
so the shape is the same either way and the reader simply has nothing to tap.

Every operation is read (transaction) → calls (**no** transaction, tens of seconds of them) → write
(transaction), for `services/stories.py`'s reason.
"""

from __future__ import annotations

import hashlib
import time
from pathlib import Path
from typing import Any, Callable

from acervo.domain.ids import now_instant
from acervo.errors import ApiError
from acervo.models import journal, load_catalogue
from acervo.models.call import audio_mime
from acervo.models.errors import ChainExhausted, ProviderError
from acervo.pronunciation import encode, speak as speaking, takes as take_store
from acervo.repository import graph, pronunciation_settings
from acervo.services import pronunciations, stories
from acervo.services.models import chain_for, refusal
from acervo.settings import Settings
from acervo.stories import narrate

NARRATE_TEMPLATE = "acervo_story_narrate"
# What tells the voice how to read a passage. Not the sentence template a word's example uses
# ("in the moment", "keep a natural pace"): this is a narrator, and the direction is a whole sentence.
SPEAK_TEMPLATE = "acervo_pronounce_story"

# The two reasons a *model's answer* is refused for being what it is rather than for the weather. A
# segmentation that fails only this way is not worth waiting for — the part is read whole instead.
UNUSABLE = frozenset({"unusable", "empty"})

# What is worth stepping over a pinned voice for: a condition that passes on its own. Written out
# here rather than imported from `work/retry.py`, because `services/` may not know jobs exist
# (`test_layering.py`) — the same three codes, decided for the same reason, in the layer that may
# say them.
PASSING = frozenset({"llm_rate_limited", "llm_unavailable", "llm_unreachable"})

Pin = tuple[str, str, "str | None"]


def narrate_story(
    settings: Settings, owner: str, device: str, story_id: str,
    gate: Callable[[], None] | None = None, progress: Callable[[int, int], None] | None = None,
) -> dict[str, Any]:
    """Record the parts that have no recording yet.

    **What is missing is re-derived here rather than passed in**, for `stories.draw_pictures`'s
    reason: a run that lost its last part to a rate limit records one part and not four. `progress`
    counts over every part, so a retry carries on from what is there and the figure never runs
    backwards.
    """
    stories._story(owner, story_id)
    parts = stories._parts(owner, story_id)
    pending = [row for row in parts if not row.get("audioSegments")]
    done = len(parts) - len(pending)
    if progress is not None:
        progress(done, len(parts))
    for row in pending:
        if gate is not None:
            gate()
        narrate_part(settings, owner, device, story_id, row["id"], gate=gate)
        done += 1
        if progress is not None:
            progress(done, len(parts))
    return {"recorded": len(pending), "parts": len(parts)}


def narrate_part(
    settings: Settings, owner: str, device: str, story_id: str, part_id: str,
    gate: Callable[[], None] | None = None,
) -> dict[str, Any]:
    """Record one part and answer with its row. The one pipeline for the job and for the route."""
    started = time.monotonic()
    story = stories._story(owner, story_id)
    parts = stories._parts(owner, story_id)
    part = next((row for row in parts if row["id"] == part_id), None)
    if part is None:
        raise ApiError(404, "not_found", "That part is not in this story.")
    language = story["language"]

    preferences = pronunciation_settings.settings(owner)
    order = preferences.order_for("stories")
    directed, every = _candidates(settings, owner, language, order)

    segments = _segments(settings, owner, part["text"], language, order, bool(directed))
    pin = _pin(parts, language)
    try:
        recordings, pin = _record(settings, owner, segments, language, order, preferences, pin,
                                  directed, every, gate)
    except ApiError as refused:
        # Every directed pair is out of reach. The part is still worth hearing, so it is read whole
        # and plainly instead — the fallback this design names — rather than left silent. One call,
        # because a part nobody can read expressively is going to be read plainly either way.
        if not directed or not every or refused.code not in PASSING:
            raise
        journal.outcome("story-audio", True, story=story_id, part=part_id,
                        result="undirected", reason=refused.code)
        segments = narrate.chunks(part["text"])
        recordings, pin = _record(settings, owner, segments, language, order, preferences, None,
                                  (), every, gate)

    row = _place_and_store(settings, owner, device, story_id, part_id, segments, recordings, pin)
    journal.outcome(
        "story-audio", False, story=story_id, part=part_id, order=order, passages=len(segments),
        directed=sum(1 for _data, said in recordings if said),
        pair=f"{pin[0]}:{pin[1]}" if pin else None, voice=pin[2] if pin else None,
        bytes=sum(len(one.get("audioRef") or "") for one in (row.get("audioSegments") or [])) and
        sum(len(data) for data, _said in recordings),
        seconds=time.monotonic() - started,
    )
    return row


def _record(settings: Settings, owner: str, segments, language: str, order: str, preferences,
            pin: Pin | None, directed, every, gate) -> tuple[list[tuple[bytes, str]], Pin | None]:
    """Say every passage, in one voice. The pair that answers the first is asked for the rest.

    **A passage already recorded is not recorded again.** A part is written only once every one of
    its passages exists, so a part that ran out of allowance halfway used to throw away the passages
    it had already paid for and buy them again on the next try — on a tier of ten calls a day, that
    is the difference between finishing a story and never finishing one. The masters go in the same
    content-addressed store a loop take uses, under the same key, so a retry costs only what is
    genuinely missing.
    """
    cache = Path(settings.takes_path)
    recordings: list[tuple[bytes, str]] = []
    for segment in segments:
        if gate is not None:
            gate()
        text = segment.text.strip()
        style = None
        if directed and segment.direction:
            style = speaking.direction(
                stories._template(settings, SPEAK_TEMPLATE), segment.direction, language)
        # A direction is recorded on the row only where one was really sent, and a directed story
        # asks only pairs that can take one — so `style` says both what was sent and what to key on.
        said = segment.direction if style else ""
        kept = _remembered(cache, text, language, style, pin, directed if style else every)
        if kept is not None:
            data, pin = kept
            recordings.append((data, said))
            continue
        spoken = _say(settings, owner, text, language, order, style,
                      preferences, pin, directed if style else every)
        answered = spoken.result.answer
        pin = (answered.provider_id, answered.model, spoken.result.voice)
        master, mime = encode.master(spoken.result.data, spoken.result.mime)
        take_store.store(cache, _take_key(text, language, style, pin), master, mime)
        recordings.append((spoken.result.data, said))
    return recordings, pin


def _take_key(text: str, language: str, direction: str | None, pin: Pin) -> str:
    return take_store.key(
        text=text, language=language, direction=direction, take=0,
        provider=pin[0], model=pin[1], voice=pin[2],
    )


def _remembered(cache: Path, text: str, language: str, style: str | None, pin: Pin | None,
                candidates) -> tuple[bytes, Pin] | None:
    """This passage, if it has already been paid for, and the voice it was paid for in.

    The story's own pair is asked for first. Failing that, every pair the order offers is tried, the
    way `pronunciations.take` probes its cache: a passage recorded yesterday by a pair that has since
    fallen out of favour is the same words in the same voice, and paying for it again on a tier of
    ten calls a day is the thing this exists to avoid. A hit becomes the pin, so the rest of the part
    is read by whoever read this.
    """
    wanted: list[Pin] = [pin] if pin else []
    for candidate in candidates:
        voice = next(iter(candidate.row.voices_for(candidate.model, language)), None)
        wanted.append((candidate.row.id, candidate.model, voice))
    for pair in wanted:
        found = take_store.find(cache, _take_key(text, language, style, pair))
        if found:
            return found[0], pair
    return None


def _say(settings: Settings, owner: str, text: str, language: str, order: str, style: str | None,
         preferences, pin: Pin | None, candidates) -> speaking.Spoken:
    """One passage, in the story's own voice where that voice will still answer.

    A pinned pair that refuses for a reason that passes — a quota, a busy provider, a dropped
    connection — is stepped over rather than waited out, and the rest of the order answers instead.
    Anything else is the deployment's own mistake and is raised, because falling through a rejected
    credential would only spend somebody else's allowance on it.
    """
    chosen = [(one.row.id, one.model) for one in candidates] or None
    if pin is not None and any(pin[:2] == pair for pair in (chosen or [])):
        try:
            return pronunciations._speak(
                settings, owner, text, language, order, style, speaking.CALLERS["stories"],
                lambda reason: None, preferences=preferences, pinned=pin,
            )
        except ApiError as refused:
            if refused.code not in PASSING:
                raise
            journal.outcome("story-audio", True, result="voice_changed", reason=refused.code,
                            pair=f"{pin[0]}:{pin[1]}")
    return pronunciations._speak(
        settings, owner, text, language, order, style, speaking.CALLERS["stories"],
        lambda reason: None, preferences=preferences, chosen=chosen,
    )


def _candidates(settings: Settings, owner: str, language: str, order: str):
    """The pairs that may read this story: the direction-capable ones, and all of them.

    Both are in the owner's own order, which is where the order is chosen; this only asks each pair
    what it can do. A pair that does not speak the language is already left out by `speakers`.
    """
    try:
        every = speaking.speakers(
            chain_for(settings, owner, pronunciations.CHAINS[order]), load_catalogue(), language)
    except ProviderError:
        return (), ()
    if order != "expressive":
        return (), every
    return tuple(one for one in every if one.row.style_for(one.model) == "instruction"), every


def _pin(parts: list[dict[str, Any]], language: str) -> Pin | None:
    """The voice this story is already being read in: the first recorded part's, if it can still be
    asked for. A pair that has since left the catalogue, lost its credential or stopped speaking the
    language is not one to insist on — a story then starts again with whatever answers first."""
    for part in sorted(parts, key=lambda row: row.get("position", 0)):
        if part.get("audioSegments") and part.get("audioProviderId") and part.get("audioModelId"):
            pair = (part["audioProviderId"], part["audioModelId"])
            try:
                usable = speaking.speakers([pair], load_catalogue(), language)
            except ProviderError:
                usable = ()
            return (*pair, part.get("audioVoice") or None) if usable else None
    return None


def _segments(settings: Settings, owner: str, text: str, language: str, order: str,
              directed: bool) -> tuple[narrate.Segment, ...]:
    """What is said, in order, and how each is directed. Joined, always exactly `text`.

    One passage and no text call at all unless a direction can actually be sent: the clear voice reads
    a part in one go by design, and a directed order with no direction-capable voice within reach is
    the same thing in practice. Asking a model where to cut a part nobody can read expressively is a
    call spent on nothing.
    """
    if order != "expressive" or not directed:
        return narrate.chunks(text)
    candidates = stories._candidates(settings, owner, "text")
    stories._require(candidates, settings, owner, "text")
    narrator = narrate.Narrator(
        load_catalogue(), candidates, stories._template(settings, NARRATE_TEMPLATE))
    try:
        tiled, usage = narrator.segment(
            narrate.build_request(language_name=speaking.language_name(language), text=text))
    except ChainExhausted as exhausted:
        if set(exhausted.reasons) <= UNUSABLE:
            # Nothing was wrong with the weather: every model was asked and none could do it. The
            # part is still worth hearing, so it is read whole, in the voice's own manner.
            journal.outcome("story-narrate", True, result="unusable", reasons=",".join(exhausted.reasons))
            return narrate.chunks(text)
        raise refusal(exhausted, "text") from None
    except ProviderError as error:
        raise refusal(error, "text") from None
    journal.outcome(
        "story-narrate", False, passages=len(tiled.segments), dropped=tiled.dropped, filled=tiled.filled,
        model=usage["model"], seconds=usage["seconds"],
    )
    return tiled.segments


def _place_and_store(
    settings: Settings, owner: str, device: str, story_id: str, part_id: str,
    segments: tuple[narrate.Segment, ...], recordings: list[tuple[bytes, str]], pin: Pin | None,
) -> dict[str, Any]:
    """Write every passage's file, then the row, then remove the files the row used to name.

    The same order `services/pronunciations._store` writes a clip in and for the same reason: a
    device that cached a passage by its reference simply misses the new one, and nothing is unlinked
    until the row naming its successor has landed.
    """
    media = Path(settings.media_path)
    written: list[dict[str, Any]] = []
    for index, (segment, (data, said)) in enumerate(zip(segments, recordings)):
        try:
            compact, mime = encode.compact(data, audio_mime(data))
        except encode.CannotEncode as unwritable:
            _discard_all(media, written, keep=())
            raise ApiError(500, "audio_unencodable",
                           f"That recording could not be stored: {unwritable}") from None
        if len(compact) > pronunciations.CLIP_LIMIT:
            _discard_all(media, written, keep=())
            raise ApiError(400, "invalid_input", "That recording is too large to keep.")
        reference = (f"stories/{story_id}/{part_id}-{index:02d}"
                     f"-{hashlib.sha256(compact).hexdigest()[:8]}.{speaking.extension_for(mime)}")
        stories._place(media, reference, compact)
        written.append({
            "text": segment.text, "direction": said, "audioRef": reference, "audioMime": mime,
            "durationSeconds": _seconds(compact),
        })

    # Read again *after* the calls, which took a while: the row this replaces may have moved — a
    # picture landed, or the other of the job and the route finished first — and a write states the
    # revision it was made from.
    fresh = graph.owned_records(owner, "storyParts", [part_id]).get(part_id)
    if fresh is None or fresh.get("deleted"):
        # The story was deleted while this was being recorded; its files go with it.
        _discard_all(media, written, keep=())
        raise ApiError(404, "not_found", "That part is not in this story any more.")
    previous = fresh.get("audioSegments") or []
    at = now_instant()
    try:
        graph.merge_graph(owner, device, {"storyParts": [{
            **fresh, "audioProviderId": pin[0] if pin else "", "audioModelId": pin[1] if pin else "",
            "audioVoice": (pin[2] or "") if pin else "", "audioSegments": written,
            "editedAt": at, "editedBy": device,
        }]}, enqueue=None)
    except Exception:
        _discard_all(media, written, keep=[one["audioRef"] for one in previous])
        raise
    _discard_all(media, previous, keep=[one["audioRef"] for one in written])
    return graph.owned_records(owner, "storyParts", [part_id])[part_id]


def _discard_all(media: Path, passages, keep) -> None:
    kept = set(keep)
    for one in passages:
        reference = one.get("audioRef") if isinstance(one, dict) else None
        if reference and reference not in kept:
            media.joinpath(reference).unlink(missing_ok=True)


def _seconds(data: bytes) -> float:
    """How long a passage lasts, so the reader can draw where it is without decoding it first."""
    import io

    try:
        import soundfile

        with soundfile.SoundFile(io.BytesIO(data)) as reading:
            return round(len(reading) / reading.samplerate, 3)
    except Exception:  # noqa: BLE001 — a length nobody could read is a length nobody needs
        return 0.0
